db/memtable_list.cc

```diff
@@ -319,12 +319,95 @@ void MemTableList::RollbackMemtableFlush(const autovector<MemTable*>& mems,
   imm_flush_needed.store(true, std::memory_order_release);
 }
 
+namespace {
+uint64_t FindMinPrepLogReferencedByMemTable(
+    VersionSet* vset, ColumnFamilyData* cfd_to_flush,
+    const autovector<MemTable*>& memtables_to_flush) {
+  uint64_t min_log = 0;
+
+  // we must look through the memtables for two phase transactions
+  // that have been committed but not yet flushed
+  for (auto loop_cfd : *vset->GetColumnFamilySet()) {
+    if (loop_cfd->IsDropped() || loop_cfd == cfd_to_flush) {
+      continue;
+    }
+
+    auto log = loop_cfd->imm()->PrecomputeMinLogContainingPrepSection(
+        memtables_to_flush);
+
+    if (log > 0 && (min_log == 0 || log < min_log)) {
+      min_log = log;
+    }
+
+    log = loop_cfd->mem()->GetMinLogContainingPrepSection();
+
+    if (log > 0 && (min_log == 0 || log < min_log)) {
+      min_log = log;
+    }
+  }
+
+  return min_log;
+}
+
+// Return the earliest log file to keep after the memtable flush is
+// finalized.
+uint64_t PrecomputeMinLogNumberToKeep(
+    ColumnFamilyData* cfd, autovector<VersionEdit*> edit_list,
+    const autovector<MemTable*>& memtables_to_flush,
+    LogsWithPrepTracker* prep_tracker, VersionSet* vset) {
+  assert(prep_tracker != nullptr);
+  // Calculate updated min_log_number_to_keep
+  // Only do it in 2PC mode, because in non-2pc mode, log number in
```

Comment:
I do not see a branch on is_2pc() here. Did you meant that the entire PrecomputeMinLogNumberToKeep should be called only in 2pc? if yes can we move this comment up to the function definition?